docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+
+5. **Client identity is tied to network address:** The Tendermint event API
+   identifies each subscriber by a (Client ID, Query) pair. In the RPC service,
+   the query is provided by the client, but the client ID is set to the TCP
+   address of the client (typically "host:port" or "ip:port").
+
+   This means that even if the server did _not_ drop subscriptions immediately
+   when the websocket connection is closed, a client may not be able to
+   reattach to its existing subscription.  Dialing a new connection is likely
+   to result in a different port (and, depending on their own proxy setup,
+   possibly a different public IP).
+
+   In isolation, this problem would be easy to work around with a new
+   subscription parameter, but it would require several other changes to the
+   handling of event subscriptions for that workaround to become useful.
+
+---
+## Decision
+
+(pending)
+
+To address the described problems, we will:
+
+1. Introduce a new API for event subscription to the Tendermint RPC service.
+   The proposed API is described in [Detailed Design](#detailed-design) below.
+
+2. This new API will target the Tendermint v0.36 release, during which the
+   current ("streaming") API will remain available as-is, but deprecated.
+
+3. The streaming API will be entirely removed in release v0.37, which will
+   require all users of event subscription to switch to the new API.
+
+> **Point for discussion:** Given that we already have ABCI++ and PBTS for
+> v0.36, it would be reasonable to suggest postponing this work until v0.37 and
+> its successor instead. The reason I proposed v0.36 for the first step is that
+> the proposed design can be added _alongside_ the existing implementation, so
+> it would not break existing usage.
+>
+> The plumbing changes would be mostly independent of the pieces that will
+> change for ABCI++ and PBTS, so working them concurrently is feasible.
+
+---
+## Detailed Design
+
+### Design Goals
+
+Specific goals of this design include:
+
+1. Remove the need for a persistent connection to each subscription client.
+   Subscribers should use the same HTTP request flow for event subscription
+   requests as for other RPC calls.
+
+2. The server retains minimal state (possibly none) per-subscriber.  In
+   particular:
+
+   - The server does not buffer unconsumed writes nor queue undelivered events
+     on a per-client basis.
+   - A client that stalls or goes idle does not cost the server any resources.
+   - Any event data that is buffered or stored is shared among _all_
+     subscribers, and is not duplicated per client.
+
+3. Slow clients have no impact (or minimal impact) on the rate of progress of
+   the consensus algorithm, beyond the ambient overhead of servicing individual
+   RPC requests.
+
+4. Clients can tell when they have missed events matching their subscription,
+   within some reasonable (configurable) window of time, and can "replay"
+   events within that window to catch up.
```

Comment:
I think you answer this further down

Reply:
I didn't answer it specifically, but yes my intent was that this would be per-node.

There is no global view of events, as far as I can tell. Arguably block events should be mostly the same, but the rest of the event stream is fairly specific to the node. So I'm not even sure what it would mean to try to coordinate this across nodes.